### ISSUE TITLE
False positive `[invalid-assignment]` error on some generator expressions

### ISSUE THREAD

**Issue by murdock-grewar** [astral-sh/ty#3222]
### Summary

`ty` appears to give a false positive `[invalid-assignment]` error on some generator expressions. A minimum reproducible example is given below:

```py
tup:  tuple[int, int] = (1, 2)
reveal_type(tuple(map(lambda n: n, tup)))
tup2: tuple[int, int] = tuple(map(lambda n: n, tup))
tup3: tuple[int, int] = tuple(n for n in tup)
```

```
error[invalid-assignment]: Object of type `tuple[Literal[1, 2], ...]` is not assignable to `tuple[int, int]`
 --> test.py:3:7
  |
1 | tup:  tuple[int, int] = (1, 2)
2 | tup2: tuple[int, int] = tuple(map(lambda n: n, tup))
3 | tup3: tuple[int, int] = tuple(n for n in tup)
  |       ---------------   ^^^^^^^^^^^^^^^^^^^^^ Incompatible value of type `tuple[Literal[1, 2], ...]`
  |       |
  |       Declared type
  |
info: rule `invalid-assignment` is enabled by default

Found 1 diagnostic
```

### Version

ty 0.0.27

**Comment by charliermarsh** [comment]
@ibraheemdev -- is this related to https://github.com/astral-sh/ruff/pull/24317?

**Comment by charliermarsh** [comment]
(Err, probably not.)

**Comment by AlexWaygood** [comment]
Thanks for the report!

For reference, pyrefly and pycroscope issue the same diagnostic as we do. Pyright, mypy and zuban are all even more strict than ty here: they all issue diagnostics for `tup2` as well as `tup3`. So no type checker has the behaviour that you're asking for here.

The reason why ty/pyrefly/pycroscope issue a diagnostic for `tup3` but not for `tup2` is because of limitations in their generics solvers. If you add a `reveal_type` call to show what the inferred type of a "standalone" expression the same as `tup2` is, we can easily see that none of those three type checkers is able to solve the type variable properly -- they all give up and infer a type of `tuple[Unknown, ...]` or `tuple[Any, ...]`:

```py
from typing import reveal_type

tup:  tuple[int, int] = (1, 2)
reveal_type(tuple(map(lambda n: n, tup)))  # pyright, mypy, zuban: `tuple[int, ...]`
                                           # ty, pyrefly: `tuple[Unknown, ...]`
                                           # pycroscope: `tuple[Any[generic_argument], ...]`
```

Assigning an object of type `tuple[Unknown, ...]` or `tuple[Any, ...]` to a variable of type `tuple[int, int]` is not an error because of a [special case for tuple assignability](https://typing.python.org/en/latest/spec/tuples.html#tuple-type-form) that the spec carves out. But assigning an object of type `tuple[int, ...]` to a variable of type `tuple[int, int]` _is_ an error, because `tuple[int, ...]` is not a subtype of `tuple[int, int]`.

Theoretically I suppose it would be possible for us to special-case the call `tuple(<generator expression that iterates over a fixed-length tuple with no filter guard>)` -- it would be safe for us to infer that this always creates a tuple of the same length as the tuple that's being iterated over there. In practice, however, that feels like very specific special-casing for a pattern that I don't think comes up _that_ much, and I don't recall seeing this problem come up before on the issue trackers of the other Python type checkers. So I think it's _probably_ not worth it for us to implement that special casing.

That said, I'd be happy for us to revisit this in the future if we see evidence that this would help other users too!

**Comment by murdock-grewar** [comment]
@AlexWaygood 
Thanks for the detailed response. I don't understand why the spec carves out that exception (I see it says _"This is useful for gradual typing."_), but it's interesting that in this case, an error is produced by availing _more_ information to the type checker.

You may like to note that the ruff rule [unnecessary-map (C417)](https://docs.astral.sh/ruff/rules/unnecessary-map/) suggests changing the expression `tuple(map(lambda n: n, tup))` into `tuple(n for n in tup)`, which actually introduces an error in this case. Although the documentation doesn't say that C417 applies to tuples, it seems that it does, at least in `ruff 0.15.6`. Ruff's fix for C417 is currently marked unsafe, but only because it may *"occasionally drop comments"*. It may be prudent to mark the fix for C417 as permanently unsafe when applied to tuples, since it introduces this typing error into ty.
